keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,634 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a corner case when Reclaim Policy is Retain and PV is referencing secrets in a certain user namespace.
+In such a case, when PVC is deleted, the PV referencing secrets remains as a resource not managed by the namespace, as a result, deletion of the secrets is blocked.
+On the other hand, provisioner secrets that are defined by using templates with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+To avoid this kind of case, an [issue](https://github.com/kubernetes-csi/external-provisioner/issues/654) for adding information on secrets in PersistentVolume as annotations is in progress. This change will allow protecting provisioner secrets from deletion even if PVC or SC for PV is deleted.
+
+## Design Details
+
+This feature can be implemented by using Lien which is being discussed in [KEP-2839](https://github.com/kubernetes/enhancements/pull/2840).
+A newly introduced Secret protection controller and a newly introduced ConfigMap protection controller update the `Liens` field of Secrets and ConfigMaps when it finds a reference to them from other resources that the controllers are watching.
+
+Note that that Secret protection controller needs to handle `VolumeSnapshotContent`, which is out-of-tree resource.
+Due to the restriction of out-of-tree resources, it needs to be handled by an external controller.
+Therefore, it is designed to make the same external controller handle both in-tree resources and out-of-tree resources, instead of creating both an in-tree controller and an external controller, separately.
+Also, ConfigMap protection controller is implemented as an external controller.
+
+To minimize the risk of unintentionally blocking the deletion, this feature is opt-in.
+Users need to add below to enable this feature per resource:
+- `kubernetes.io/enable-secret-protection` annotation to Secret
+- `kubernetes.io/enable-configmap-protection` annotation to ConfigMap
+
+Basic flows of Secret protection controller and ConfigMap protection controller are as follows:
+
+- Secret protection controller:
+  - The controller watches `Pod`, `PersistentVolume`, and `VolumeSnapshotContent`
+    - Create/Update event:
+      - For all the `Secret`s referenced by the resource:
+        - If the `Secret` doesn't have `kubernetes.io/enable-secret-protection: "yes"` annotation, do nothing
+        - Update the `Secret`'s `Liens` field to add `kubernetes.io/secret-protection ${version}/${kind}/${namespace}/${name}`.
```

Comment:
This design is not going to scale.
A secret may be referenced from thousands of pods and we would end-up with thousands of liens?
At some point, we simply hit the size limit for the object itself.

We shouldn't put metadata of all referencing objects here...

Reply:
I understand that we should consider scalablity. Let's discuss it on [this thread](https://github.com/kubernetes/enhancements/pull/2840#discussion_r703066831) in #2840.

Reply:
Added #2840 as a non-goal for this KEP.

Reply:
Updated "Design Details" section to utilize the per-controller version of Lien.